### PR TITLE
Add command for changing expression value

### PR DESCRIPTION
**Pull request by 4e6**
### Pull Request Description

Implement `text/applyExpressionValue` command.

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] ~If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide dist` and `./run ide watch`.~

### CODE REVIEW

File: engine/runtime/src/main/scala/org/enso/compiler/context/ChangesetBuilder.scala [hunk 1]
```diff
@@ -58,34 +58,26 @@ final class ChangesetBuilder[A: TextEditor: IndexedSource](
     * @return the computed changeset
     */
   @throws[CompilerError]
-  def build(edits: Seq[TextEdit]): Changeset[A] = {
-
-    val simpleEditOption: Option[TextEdit] = edits match {
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/scala/org/enso/compiler/context/ChangesetBuilder.scala:63]: Removing this removes the only way to test the behavior with real IDE that we had so far. Wouldn't it be better to keep this `simpleEditOption` logic in? It hurts nothing, right?

File: engine/runtime/src/main/java/org/enso/interpreter/service/ExecutionService.java [hunk 2]
```diff
@@ -4,25 +4,10 @@
 import com.oracle.truffle.api.TruffleLogger;
 import com.oracle.truffle.api.instrumentation.EventBinding;
 import com.oracle.truffle.api.instrumentation.ExecutionEventListener;
-import com.oracle.truffle.api.interop.ArityException;
-import com.oracle.truffle.api.interop.InteropLibrary;
-import com.oracle.truffle.api.interop.UnknownIdentifierException;
-import com.oracle.truffle.api.interop.UnsupportedMessageException;
-import com.oracle.truffle.api.interop.UnsupportedTypeException;
+import com.oracle.truffle.api.interop.*;
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/interpreter/service/ExecutionService.java]: I'd rather avoid using `*` imports. Would it be possible to configure our `sbt javafmt` to reject star imports, @hubertp?

**Review by JaroslavTulach — APPROVED** [review]

File: engine/runtime-with-instruments/src/test/java/org/enso/interpreter/test/instrument/IncrementalUpdatesTest.java [hunk 3]
```diff
@@ -227,17 +228,18 @@ private static String extractPositions(String code, String chars, Map<Character,
     return executionCompleteEvents;
   }
 
-  private List<Runtime$Api$Response> sendEditFile(String originalText, String newText) {
+  private List<Runtime$Api$Response> sendExpressionValue(String originalText, String newText) {
     assertNotNull("Main file must be defined before", mainFile);
-    context.send(Request(new Runtime$Api$EditFileNotification(
+    context.send(Request(new Runtime$Api$SetExpressionValueNotification(
```
Comment (JaroslavTulach) [inline engine/runtime-with-instruments/src/test/java/org/enso/interpreter/test/instrument/IncrementalUpdatesTest.java:273]: The test could be _duplicated_. As we can still recognize the simple `EditFileNotification` - we could have two variants of the test - one sending `SetExpressionValueNotification` and the other one `EditFileNotification`.

File: docs/language-server/protocol-language-server.md [hunk 4]
```diff
@@ -2823,14 +2823,8 @@ interface TextApplyExpressionValue {
...
-  /** Edited file. */
-  path: Path;
-
-  /** The SHA hash of the file before setting the new value to the node. */
-  oldVersion: SHA3-224;
-
-  /** The SHA hash of the file after setting the new value to the node. */
-  newVersion: SHA3-224;
+  /** The file edit. */
+  edit: FileEdit;
```
Comment (MichaelMauderer) [inline docs/language-server/protocol-language-server.md]: How is this file edit applied? This seems to allow this API to contain arbitrary edits in addition to the expression update.

Comment (4e6) [reply docs/language-server/protocol-language-server.md]: The edit is applied as-is. There are two choices
- We can take the expression value from the edit, but then IDE should guarantee that the edit fully replaces previous value, i.e. partial edits, like `100` -> `10` by removing the `0` are not allowed. It should be an edit setting value `10`.
- We can allow partial edits, and additionally set the new value in the `expressionValue` field (this approach)

Comment (MichaelMauderer) [reply docs/language-server/protocol-language-server.md]: Doesn't this again allow inconsistent data to be sent? Let's say that the `expressionValue` says the value is `100`, but the `FileEdit` adds the value `200`? 

Is there any benefit to allowing additional edits or partial edits as opposed to just replacing the `expressionValue` as a whole?

Comment (4e6) [reply docs/language-server/protocol-language-server.md]: Hm, I see. Ok, if it works for IDE I'll leave that flat structure.

**Comment by 4e6** [comment]
@MichaelMauderer I updated the thing and removed the `expressionValue` field. The expression value is taken from the `TextEdit`.

File: engine/runtime/src/main/java/org/enso/interpreter/service/ExecutionService.java (hunk 2 above)
Comment (hubertp) [reply engine/runtime/src/main/java/org/enso/interpreter/service/ExecutionService.java]: google-java-formatter (called by javafmt behind the scenes) does not really support that and it is out-of-scope IIRC

**Comment by MichaelMauderer** [comment]
> @MichaelMauderer I updated the thing and removed the `expressionValue` field. The expression value is taken from the `TextEdit`.

But aren't we now back to the original discussion we had on the design of the API? What happens if an arbitrary edit in `TextEdit` changes something else in here?

**Comment by 4e6** [comment]
> But aren't we now back to the original discussion we had on the design of the API? What happens if an arbitrary edit in `TextEdit` changes something else in here?

The program will most likely fail during the execution because a new value for the expression is taken from the `TextEdit`.

It is a change in the program, and `TextEdit` is required anyway (to keep the text buffer up-to-date). We can't make it safer because `TextEdit` operates on a text-level, so you don't know what it changes in the program until you apply the change and re-compile the program (and we want to avoid that)

**Review by MichaelMauderer — APPROVED** [review]
After some discussion, we found that the current proposal is the best we can currently do that also achieves the desired performance goals.

**Pull request merged**